sdk/helper/scaleutils/nodeselector/least_busy.go

```diff
@@ -0,0 +1,199 @@
+// for entry into the heap. Any error here is terminal and means we are unable
+// to safely understand the least allocated nodes.
+func (l *leastBusyClusterScaleInNodeSelector) computeNodeResources(node *api.NodeListStub) (*nodeResourceStats, error) {
+
+	nodeAllocs, _, err := l.client.Nodes().Allocations(node.ID, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	allocated := computeNodeAllocatedResources(nodeAllocs)
+	total := computeNodeTotalResources(node)
+
+	mem := computeNodeAllocatedPercentage(allocated.Memory.MemoryMB, total.Memory.MemoryMB)
+	cpu := computeNodeAllocatedPercentage(allocated.Cpu.CpuShares, total.Cpu.CpuShares)
+
+	return &nodeResourceStats{
+		consumed: math.Ceil((mem + cpu) / 2),
+		node:     node,
+	}, nil
+}
+
+// computeNodeAllocatedResources calculates the amount of allocated resource on
+// the node as used by running allocations.
+func computeNodeAllocatedResources(allocs []*api.Allocation) *api.NodeResources {
+	var mem, cpu int
+
+	for _, alloc := range allocs {
+		if alloc.ClientStatus != api.AllocClientStatusRunning {
+			continue
+		}
+		mem += *alloc.Resources.MemoryMB
+		cpu += *alloc.Resources.CPU
+	}
+
+	return &api.NodeResources{
+		Cpu:    api.NodeCpuResources{CpuShares: int64(cpu)},
+		Memory: api.NodeMemoryResources{MemoryMB: int64(mem)},
+	}
+}
+
+// computeNodeTotalResources calculates the node allocatable resources and
+// takes into account reserved resources.
+func computeNodeTotalResources(node *api.NodeListStub) *api.NodeResources {
+	total := api.NodeResources{}
+
+	r := node.NodeResources
+	res := node.ReservedResources
+	if res == nil {
+		res = &api.NodeReservedResources{
+			Cpu:    api.NodeReservedCpuResources{},
+			Memory: api.NodeReservedMemoryResources{},
+		}
+	}
+	total.Cpu.CpuShares = r.Cpu.CpuShares - int64(res.Cpu.CpuShares)
+	total.Memory.MemoryMB = r.Memory.MemoryMB - int64(res.Memory.MemoryMB)
+	return &total
+}
+
+// computeNodeAllocatedPercentage helps calculate the percentage allocation of
+// a node resource.
+func computeNodeAllocatedPercentage(allocated, total int64) float64 {
+	return math.Ceil((float64(allocated) * float64(100)) / float64(total))
+}
+
+// nodeResourceStats encapsulates a single nodes resource consumption so that
+// we can correctly perform our calculations to find the least allocated nodes
+// for termination.
+type nodeResourceStats struct {
+
+	// consumed is the total percentage of memory and CPU that is consumed on
+	// the node. This provides the basis for node comparison.
+	consumed float64
+
+	// node is a pointer to the node which this heap entry represents. It
+	// provides an easy way to pop entries in the way we need to return a list
+	// to the caller.
+	node *api.NodeListStub
+
+	// index tracks the index on the heap that this entry currently occupies.
+	index int
+}
+
+// nodeResourceConsumption is our container/heap implementation.
+type nodeResourceConsumption []*nodeResourceStats
```

Comment:
I don't really have a strong opinion either way. That being said, I don't see a possibility for nil pointer panics in the current code and from other examples I have seen using a pointer seems common place.

Reply:
It doesn't happen right now because you don't use `heap.Push`, but as soon as you do you would get a panic:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x12cb98f]

goroutine 82 [running]:
github.com/hashicorp/nomad-autoscaler/sdk/helper/scaleutils/nodeselector.nodeResourceConsumption.Less(...)
        /Users/laoqui/go/src/github.com/hashicorp/nomad-autoscaler/sdk/helper/scaleutils/nodeselector/least_busy.go:170
container/heap.down(0x13f5ea0, 0xc00000e040, 0x31, 0x64, 0x12ee0a0)
        /nix/store/iadyfrad1x7c05frqsqgw60z6mkjf9wl-go-1.14.8/share/go/src/container/heap/heap.go:112 +0xd3
container/heap.Init(0x13f5ea0, 0xc00000e040)
        /nix/store/iadyfrad1x7c05frqsqgw60z6mkjf9wl-go-1.14.8/share/go/src/container/heap/heap.go:46 +0x78
github.com/hashicorp/nomad-autoscaler/sdk/helper/scaleutils/nodeselector.BenchmarkSelectHeap2(0xc000014380)
        /Users/laoqui/go/src/github.com/hashicorp/nomad-autoscaler/sdk/helper/scaleutils/nodeselector/least_busy_test.go:261 +0x1e6
testing.(*B).runN(0xc000014380, 0x64)
        /nix/store/iadyfrad1x7c05frqsqgw60z6mkjf9wl-go-1.14.8/share/go/src/testing/benchmark.go:191 +0xe8
testing.(*B).launch(0xc000014380)
        /nix/store/iadyfrad1x7c05frqsqgw60z6mkjf9wl-go-1.14.8/share/go/src/testing/benchmark.go:321 +0xea
created by testing.(*B).doBench
        /nix/store/iadyfrad1x7c05frqsqgw60z6mkjf9wl-go-1.14.8/share/go/src/testing/benchmark.go:276 +0x55
exit status 2
FAIL    github.com/hashicorp/nomad-autoscaler/sdk/helper/scaleutils/nodeselector        12.671s
FAIL
```

Reply:
Resolving this issue. After some internal discussion and benchmarking we are moving from heap to sort.